rust/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+#![no_std]
+#![warn(missing_docs)]
+
+//! # GLOME - Generic Low-Overhead Message Exchange
+//!
+//! GLOME is a lightweight message authentication protocol based on X25519
+//! keys and HMAC-SHA256. See <https://github.com/google/glome> for details on
+//! the protocol, usage patterns and implementations in other languages.
+//!
+//! The Rust implementation of GLOME works with its own [PrivateKey] and
+//! [PublicKey] traits to support more than one backing cryptography crate.
+//! It aims to provide implementations for the crates most commonly used, which
+//! can be activated with the corresponding crate feature. The default and
+//! recommended setting is to use [x25519_dalek]. Implementations should be
+//! verified with the test vectors in the `tests` module.
+
+use hmac::{Hmac, Mac};
+use sha2::Sha256;
+
+type HmacSha256 = Hmac<Sha256>;
+
+/// An X25519 public key.
+pub trait PublicKey {
+    /// to_bytes encodes the public key as a byte array according to RFC 7748.
+    fn to_bytes(&self) -> [u8; 32];
+}
+
+/// An X25519 private key.
+pub trait PrivateKey {
+    /// PublicKey is the type of public keys corresponding to this type.
+    type PublicKey: PublicKey;
+
+    /// dh corresponds to the x25519 function from RFC 7748, computing a shared secret from a
+    /// private key and a peer key.
+    fn dh(&self, theirs: &Self::PublicKey) -> [u8; 32];
+
+    /// public_key computes the public key corresponding to this private key.
+    fn public_key(&self) -> Self::PublicKey;
+}
+
+/// Compute a GLOME tag.
+///
+/// The message counter argument `ctr` is used to prevent replay attacks in a
+/// series of messages. If only a single message is exchanged between two key
+/// pairs, it can be set to `0u8`. The counter value is not secret, but needs
+/// to be integrity protected. Usually, this is accomplished by both parties
+/// counting sent and received messages internally.
+///
+/// The returned tag can be verified by the
+/// other party, using [verify] with its private key, our public key and the
+/// counter.
+pub fn tag<T: PrivateKey>(ours: &T, theirs: &T::PublicKey, ctr: u8, msg: &[u8]) -> [u8; 32] {
+    let key = [
+        ours.dh(theirs),
+        theirs.to_bytes(),
+        ours.public_key().to_bytes(),
+    ]
+    .concat();
+
+    HmacSha256::new_from_slice(&key)
+        .expect("HMAC can take key of any size")
+        .chain_update(&[ctr])
```

Comment:
`cargo clippy` warning:

```suggestion
        .chain_update([ctr])
```